[sup] Enable client-authentication with TLS

stevendanna opened the pull request:
TLS allows clients to also provide a certificate when initiating a
connection. This certifcate can be used to authenticate clients as an
alternative to other HTTP authentication schems such as basic auth.

I've factored out the TLS configuration into a struct because passing
around `Option<(PathBuf, PathBuf, Option<PathBuf>)>` felt like it
would become a bit error prone and hard to extend.

Signed-off-by: Steven Danna <steve@chef.io>

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

baumanj commented on components/sup/src/manager/mod.rs:
```diff
@@ -1239,29 +1248,42 @@ impl Manager {
+            if added < 1 {
+                return Err(sup_error!(Error::InvalidCertFile(path.to_path_buf())));
+            }
+            AllowAnyAuthenticatedClient::new(root_store)
```
We can reduce the duplication here with `and_then`:
```rust
            root_store
                .add_pem_file(ca_file)
                .and_then(|(added, _)| {
                    if added < 1 {
                        Err(())
                    } else {
                        Ok(AllowAnyAuthenticatedClient::new(root_store))
                    }
                })
                .map_err(|_| sup_error!(Error::InvalidCertFile(path.clone())))?
```

stevendanna commented:
Updated the style issues and changed the flag requirements so that you get an error if you try to turn on client-auth but haven't turned on the other tls flags.  To be honest, it will be a while before we have time to figure out how to transition to turning this on so no rush.

baumanj commented:
Our merge freeze is lifted now. Let me know when you've rebased to fix the conflict, and I can get this merged, @stevendanna

stevendanna commented:
Rebased and re-rustfmt'd

baumanj commented:
Thanks again!
